File: Mediation/MoPubExample/Objective-C/VponMopubSample_objC/VponMopubSampleObjC/Mopub/VponMopubNativeViewController.m
//
//  VponMopubNativeViewController.m
//  VponMopubSampleObjC
//
//  Created by EricChien on 2017/6/13.
//  Copyright © 2017年 Soul. All rights reserved.
//

#import "VponMopubNativeViewController.h"

#import <MoPub.h>
#import "MPVponNativeCustomEvent.h"

#import "MPVponNativeAdView.h"

@interface VponMopubNativeViewController () <MPNativeAdDelegate>

@property (weak, nonatomic) IBOutlet UIButton *requestButton;

@property (weak, nonatomic) IBOutlet UIView *loadNativeView;

@property (strong, nonatomic) MPNativeAd *mpNativeAd;

@end

@implementation VponMopubNativeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"MoPub - Native";
    [self requestButtonDidTouch:self.requestButton];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Button Method

- (IBAction)requestButtonDidTouch:(UIButton *)sender {
    
    sender.enabled = NO;
    
    MPStaticNativeAdRendererSettings *settings = [[MPStaticNativeAdRendererSettings alloc] init];
    #warning create a custom view for your business
    settings.renderingViewClass = [MPVponNativeAdView class];
    
    MPNativeAdRendererConfiguration *config = [MPStaticNativeAdRenderer rendererConfigurationWithRendererSettings:settings];
    config.supportedCustomEvents = @[@"MPVponNativeCustomEvent"];
    
    #warning set ad unit id
    MPNativeAdRequest *adRequest = [MPNativeAdRequest requestWithAdUnitIdentifier:@"ddeaeccecddf44cea7732465a9138239" rendererConfigurations:@[config]];
    
    MPNativeAdRequestTargeting *targeting = [MPNativeAdRequestTargeting targeting];
    targeting.desiredAssets = [NSSet setWithObjects:kAdTitleKey, kAdTextKey, kAdCTATextKey, kAdIconImageKey, kAdMainImageKey, kAdStarRatingKey, nil];
//    targeting.localExtras = @{
//        @"contentURL": @"https://www.vpon.com",
//        @"contentData": @{@"key1": @"Mopub", @"key2": @(1.2), @"key3": @(YES)},
//        @"friendlyObstructions": @[@{ @"view": [[UIView alloc] init], @"purpose": @(2), @"desc": @"not_visible"}]
//    };
    adRequest.targeting = targeting;
    
    __block typeof(self) weakSelf = self;
    [adRequest startWithCompletionHandler:^(MPNativeAdRequest *request, MPNativeAd *response, NSError *error) {
        if (error) {
            NSLog(@"Failed to receive native with error : %@", error.localizedFailureReason);
        } else {
            weakSelf.mpNativeAd = response;
            weakSelf.mpNativeAd.delegate = weakSelf;
            UIView *nativeAdView = [response retrieveAdViewWithError:nil];
            nativeAdView.frame = weakSelf.loadNativeView.bounds;
            [weakSelf.loadNativeView addSubview:nativeAdView];
        }
        weakSelf.requestButton.enabled = YES;
    }];
}

#pragma mark MPNativeAd Delegate

-(UIViewController *)viewControllerForPresentingModalView {
    return self;
}

- (void)willPresentModalForNativeAd:(MPNativeAd *)nativeAd {
    NSLog(@"willPresentModalForNativeAd");
}

- (void)didDismissModalForNativeAd:(MPNativeAd *)nativeAd {
    NSLog(@"didDismissModalForNativeAd");
}

- (void)willLeaveApplicationFromNativeAd:(MPNativeAd *)nativeAd {
    NSLog(@"willLeaveApplicationFromNativeAd");
}

@end

File: Mediation/MoPubExample/Swift/VponMopubSample_swift/VponMopubSampleSwift/Mopub/VponMopubNativeViewController.swift
//
//  VponMopubNativeViewController.swift
//  VponMopubSampleSwift
//
//  Created by EricChien on 2017/6/13.
//  Copyright © 2017年 Soul. All rights reserved.
//

import UIKit

import MoPub

class VponMopubNativeViewController: UIViewController {
    
    @IBOutlet weak var requestButton: UIButton!
    
    @IBOutlet weak var loadNativeView: UIView!
    
    var mpNativeAd: MPNativeAd!

    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = "MoPub - Native"
        requestButtonDidTouch(requestButton)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    // MARK: Button Method
    
    @IBAction func requestButtonDidTouch(_ sender: UIButton) {
        
        sender.isEnabled = false
        
        let settings = MPStaticNativeAdRendererSettings.init()
        settings.renderingViewClass = MPVponNativeAdView.classForCoder()
        
        let config = MPStaticNativeAdRenderer.rendererConfiguration(with: settings)!
        config.supportedCustomEvents = ["MPVponNativeCustomEvent"]
        
        // TODO: set ad unit id
        let adRequest = MPNativeAdRequest(adUnitIdentifier: "ddeaeccecddf44cea7732465a9138239", rendererConfigurations: [config])
        
        let targeting = MPNativeAdRequestTargeting()
        targeting?.desiredAssets = NSSet(objects: kAdTitleKey, kAdTextKey, kAdCTATextKey, kAdIconImageKey, kAdMainImageKey, kAdStarRatingKey) as? Set<AnyHashable>
        targeting?.localExtras = ["contentURL":"https://www.vpon.com", "contentData": ["key1": "Mopub", "key2": 1.2, "key3": true]]
        
        adRequest?.targeting = targeting
        adRequest?.start { (request, response, error) in
            if error != nil {
                print("Failed to receive banner")
            } else {
                self.mpNativeAd = response
                self.mpNativeAd.delegate = self
                do {
                    let nativeAdView = try response?.retrieveAdView()
                    nativeAdView?.frame = self.loadNativeView.bounds
                    self.loadNativeView.addSubview(nativeAdView!)
                } catch {
                    print(error)
                }
            }
        }
    }
}

extension VponMopubNativeViewController: MPNativeAdDelegate {
    
    func willPresentModal(for nativeAd: MPNativeAd!) {
        print("will Present Modal For NativeAd");
    }
    
    func didDismissModal(for nativeAd: MPNativeAd!) {
        print("did Dismiss Modal For NativeAd")
    }
    
    func willLeaveApplication(from nativeAd: MPNativeAd!) {
        print("will Leave Application From NativeAd")
    }
    
    func viewControllerForPresentingModalView() -> UIViewController! {
        return self
    }
}
